api-helpers/contentCreatorFunctions.ts

```diff
@@ -0,0 +1,175 @@
+import { URL } from 'url';
+
+import Boom from '@hapi/boom';
+import cheerio from 'cheerio';
+import Slugify from 'slugify';
+
+import { closeConnection, openConnection } from './db';
+
+const NEVER = new Date(0);
+
+type BlogData = {
+  readonly name: string;
+  readonly href: string;
+  readonly favicon: string;
+  readonly rss: string;
+};
+
+type Feed = {
+  readonly type: string;
+  readonly url: string;
+  readonly title: string;
+};
+
+export const addContentCreator = async (url: string) => {
+  try {
+    const prisma = await openConnection();
+    const blogData = await getBlogData(url);
+    return await prisma.blog.create({
+      data: { ...blogData, lastUpdateDate: NEVER, slug: Slugify(blogData.name, { lower: true }) },
+    });
+  } finally {
+    await closeConnection();
+  }
+};
+
+const getBlogData = async (url: string): Promise<BlogData> => {
+  const youtubeRss = getYoutubeRss(url);
+  if (youtubeRss) {
+    const response = await fetch(youtubeRss);
+    const xmlText = await response.text();
+    if (xmlText) {
+      const $ = cheerio.load(xmlText);
+      return {
+        name: getBlogName($),
+        href: url,
+        favicon:
+          (await getYoutubeChannelImageSource(url)) ||
+          'https://www.youtube.com/s/desktop/d743f786/img/favicon_48.png',
+        rss: youtubeRss,
+      };
+    }
+    throw Boom.badData();
+  }
+  const response = await fetch(url);
+  const htmlText = await response.text();
+  if (htmlText) {
+    const $ = cheerio.load(htmlText);
+    const feeds = searchFeed(url, $);
+    const rssFeedUrl = feeds[0].url;
+    const blogData = await getBlogDataFromRss(rssFeedUrl);
+    if (blogData) {
+      return {
+        name: blogData.name,
+        href: url,
+        favicon: blogData.favicon,
+        rss: rssFeedUrl,
+      };
+    }
+  }
+
+  throw Boom.badData();
+};
+
+const getYoutubeRss = (url: string) => {
+  const regex = /^(http(s)?:\/\/)?((w){3}.)?youtu(be|.be)?(\.com)?\/.+/gm;
+  const isYoutubeUrl = regex.test(url);
+
+  if (isYoutubeUrl) {
+    return getYouTubeChannelFeedUrl(url);
+  }
+  return undefined;
+};
+
+const getYouTubeChannelFeedUrl = (url: string) => {
+  if (getYouTubeChannelIdFromUrl(url)) {
+    return `https://www.youtube.com/feeds/videos.xml?channel_id=${getYouTubeChannelIdFromUrl(url)}`;
+  }
+  if (getYouTubeUserIdFromUrl(url)) {
+    return `https://www.youtube.com/feeds/videos.xml?user=${getYouTubeUserIdFromUrl(url)}`;
+  }
+  return undefined;
+};
+
+const getYouTubeChannelIdFromUrl = (url: string) => {
+  const urlWithoutParams = getUrlWithoutParams(url);
+  return urlWithoutParams.split('channel/')[1].split('/')[0];
+};
+
+const getYouTubeUserIdFromUrl = (url: string) => {
+  const urlWithoutParams = getUrlWithoutParams(url);
+  return urlWithoutParams.split('user/')[1].split('/')[0];
+};
+
+const getUrlWithoutParams = (url: string) => {
+  return url.split('?')[0];
+};
```

Comment:
W przypadku tych dwóch funkcji można by skorzystać z modułu URL Node'a

Reply:
Można, aczkolwiek nie wiem, czy to poprawi czytelność tej funkcji.